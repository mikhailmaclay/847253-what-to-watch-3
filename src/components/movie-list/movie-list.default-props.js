import {Config} from '../../consts';

export default {
  renderedMovieCards: Config.MOVIE_CARDS_COUNT_AT_START // withCounter
};
