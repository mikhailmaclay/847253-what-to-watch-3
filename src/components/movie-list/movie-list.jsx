// Libraries
import React from 'react';
// PropTypes
import propTypes from './movie-list.prop-types';
// Constants and utils
import {Config} from '../../consts';
import createHOC from '../../utils/components/create-hoc';
// Components
import MovieCardWrapped from '../movie-card/movie-card';
import Button from '../button/button';

function MovieList({movies, /* withCounter: */ isWithRenderedMovieCardsCounter, withCounter}) {
  const renderMovieCards = (count) => (
    movies.slice(0, count).map((movie) => <MovieCardWrapped key={movie.id} {...movie}/>)
  );

  if (!isWithRenderedMovieCardsCounter) { // if not a withCounter HOC
    return (
      <div className="catalog__movie-list">
        {renderMovieCards(Config.MOVIE_CARDS_COUNT_AT_START)}
      </div>
    );
  }

  const handleButtonClick = () => {
    withCounter.renderedMovieCards.increment(Config.MOVIE_CARDS_COUNT_BY_LOAD);
  };

  return (
    <>
      <div className="catalog__movie-list">
        {renderMovieCards(withCounter.renderedMovieCards.value)}
      </div>
      {withCounter.renderedMovieCards.value < movies.length &&
        <div className="catalog__more">
          <Button className="catalog__button" onClick={handleButtonClick}>Show more</Button>
        </div>
      }
    </>
  );
}

MovieList.propTypes = propTypes;

const MovieListMemo = React.memo(MovieList);
const MovieListWrapped = createHOC(`withCounter`, {counterName: `renderedMovieCards`, initialValue: Config.MOVIE_CARDS_COUNT_AT_START, isResettableOnUpdate: true})(MovieListMemo);

export default MovieListWrapped;
export {MovieList, MovieListMemo};
