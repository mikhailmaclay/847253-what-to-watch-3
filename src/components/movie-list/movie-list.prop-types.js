import PropTypes from 'prop-types';

export default {
  movies: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired
  })).isRequired,
  count: PropTypes.number.isRequired,
  incrementCounter: PropTypes.func.isRequired,
  onMouseOver: PropTypes.func
};
