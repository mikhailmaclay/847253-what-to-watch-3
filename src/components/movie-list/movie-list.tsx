// Libraries
import * as React from 'react';
// Constants and utils
import {Config} from '../../constants/consts';
import createHOC from '../../utils/components/create-hoc';
// Components
import MovieCardWrapped from '../movie-card/movie-card';
import Button from '../button/button';
//
import {Movie} from '../../types';

interface Props {
  movies: Movie[];
  withCounter?: {
    renderedMovieCards: {
      value: number;
      increment: (value: number) => void;
    };
  };
}

function MovieList(props: Props) {
  const {movies, withCounter} = props;

  const renderMovieCards = (count) => (
    movies.slice(0, count).map((movie) => <MovieCardWrapped key={movie.id} {...movie}/>)
  );

  if (!withCounter) { // if not a withCounter HOC
    return (
      <div className="catalog__movie-list">
        {renderMovieCards(Config.MOVIE_CARDS_COUNT_AT_START)}
      </div>
    );
  }

  const handleButtonClick = () => {
    withCounter.renderedMovieCards.increment(Config.MOVIE_CARDS_COUNT_BY_LOAD);
  };

  return (
    <>
      <div className="catalog__movie-list">
        {renderMovieCards(withCounter.renderedMovieCards.value)}
      </div>
      {withCounter.renderedMovieCards.value < movies.length &&
        <div className="catalog__more">
          <Button className="catalog__button" onClick={handleButtonClick}>Show more</Button>
        </div>
      }
    </>
  );
}

const MovieListMemo = React.memo(MovieList);
const MovieListWrapped = createHOC(`withCounter`, {counterName: `renderedMovieCards`, initialValue: Config.MOVIE_CARDS_COUNT_AT_START, isResettableOnUpdate: true})(MovieListMemo);

export default MovieListWrapped;
export {MovieList, MovieListMemo};
