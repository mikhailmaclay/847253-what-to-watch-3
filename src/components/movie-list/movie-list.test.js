// Libraries
import React from 'react';
import renderer from 'react-test-renderer';
import {BrowserRouter} from 'react-router-dom';
// Components
import {MovieList} from './movie-list';

const movies = [
  {
    id: 1,
    name: `Fantastic Beasts: The Crimes of Grindelwald`,
    images: [`/img/movies/images/2-1.jpg`],
  }
];

describe(`<MovieList/>`, () => {
  const props = {
    movies
  };

  it(`should render correctly`, () => {
    const result = renderer.create(
        <BrowserRouter>
          <MovieList {...props}/>
        </BrowserRouter>
    ).toJSON();

    expect(result).toMatchSnapshot();
  });
});
