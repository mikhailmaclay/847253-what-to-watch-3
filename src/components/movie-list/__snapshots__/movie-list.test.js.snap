// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieList/> should render correctly 1`] = `
<div
  className="catalog__movie-list"
>
  <article
    className="small-movie-card catalog__movie-card"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <a
      className="small-movie-card__image"
      href="/films/1"
      onClick={[Function]}
    >
      <img
        alt="Fantastic Beasts: The Crimes of Grindelwald"
        height="175"
        src="/img/movies/images/2-1.jpg"
        width="280"
      />
    </a>
    <h3
      className="small-movie-card__title"
    >
      <span
        className="small-movie-card__link"
      >
        Fantastic Beasts: The Crimes of Grindelwald
      </span>
    </h3>
  </article>
</div>
`;
